File: Grind75/63-MinimumHeightTrees/cpp/MinimumHeightTrees.cpp
#include "MinimumHeightTrees.h"
#include "prettyPrint.h"

using namespace std;

void longestPath(const int root, const vector<vector<int>> &adj,
    vector<int> &path, vector<int> &longest) {
  // cout << "root: " << root << endl;
  path.push_back(root);
  bool noNeighbors = true;
  for (const int &nbr : adj[root]) {
    if (path.size() > 1 && nbr == path[path.size()-2])
      continue;
    longestPath(nbr, adj, path, longest);
    noNeighbors = false;
    // cout << "nbr: " << nbr << ", path: " << prettyString(nbrPath) << endl;
  }
  if ( (noNeighbors && longest.empty())
      || (!longest.empty() && path.size() > longest.size()) )
    longest = path;
  path.pop_back();
}

vector<int> Solution::findMinHeightTrees(int n, vector<vector<int>>& edges) {
  // Build adjacency list
  vector<vector<int>> adj(n);
  for (const vector<int> &edge : edges) {
    adj[edge[0]].push_back(edge[1]);
    adj[edge[1]].push_back(edge[0]);
  }

  vector<int> path, pathFrom0, pathFromFurthest;
  longestPath(0, adj, path, pathFrom0);
  longestPath(pathFrom0.back(), adj, path, pathFromFurthest);
  prettyPrint(pathFromFurthest);
  int N = pathFromFurthest.size();
  if (pathFromFurthest.size() % 2 == 0)
    return { pathFromFurthest[(N-1)/2], pathFromFurthest[(N-1)/2 + 1] };
  else
    return { pathFromFurthest[(N-1)/2] };
}

File: Grind75/63-MinimumHeightTrees/python/MinimumHeightTrees.py
#!/bin/python3

from __future__ import annotations
from collections import defaultdict, deque
import sys
from typing import DefaultDict, Deque, Dict
sys.path.append('../../Utility/python')
from PrettyPrint import *
from Colors import *

class Solution:
  def findMinHeightTrees(self, n: int, edges: List[List[int]]) -> List[int]:
    adj: List[List[int]] = [ [] for _ in range(n) ]
    for v1, v2 in edges:
      adj[v1].append(v2)
      adj[v2].append(v1)

    leaves: Deque[int] = deque()
    degrees: List[int] = [0] * n
    for node, neighbors in enumerate(adj):
      degrees[node] = len(neighbors)
      if len(neighbors) == 1:
        leaves.append(node)

    while leaves:
      if n <= 2:
        return list(leaves)
      for _ in range(len(leaves)):
        node = leaves.popleft()
        n -= 1
        for nbr in adj[node]:
          degrees[nbr] -= 1
          if degrees[nbr] == 1:
            leaves.append(nbr)
    
    return [0]


if __name__ == '__main__':
  soln = Solution()
  testCases = [
    ( 1, [], [0]),
    ( 2, [[0,1]], [0,1] ),
    ( 3, [[0,1],[1,2]], [1] ),
    ( 4, [[0,1],[1,2],[2,3]], [1,2] ),
    ( 4, [[1,0],[1,2],[1,3]], [1] ),
    ( 5, [[0,1],[1,2],[2,3],[2,4]], [1,2] ),
  ]

  for n, edges, answer in testCases:
    colorPrint(LIGHT_BLUE, f'{n}:', prettyString2D(edges, outSep=', '),
               "=>", answer, end='')
    result = soln.findMinHeightTrees(n, edges)
    print(' --> ', end='')
    colorPrint(GREEN if set(result) == set(answer) else RED, result)

